index.bs

```diff
@@ -2015,14 +1903,67 @@ Issue: This needs to be generalised to work with realms too.
 <div algorithm="remote end steps for session.unsubscribe">
 The [=remote end steps=] with |session| and |command parameters| are:
 
-1. Let the |list of event names| be the value of the <code>events</code> field of
-   |command parameters|.
+1. If |command parameters| matches the <code>session.SubscriptionRequest</code> production:
+
+   1. Let the |event names| be the value of the <code>events</code> field of
+      |command parameters|.
+
+   1. Let the |contexts| be the value of the <code>contexts</code>
+      field of |command parameters| if it is present or an empty [=/set=] if it isn't.
+
+   1. Let |new subscriptions| to be a [=/list=].
+
+   1. For each |subscription| of |session|'s [=subscriptions=]:
+
+      1. If [=set/intersection=] of |subscription|'s event names and |event names| is an empty [=/set=],
+         continue.
+
+      1. If |contexts| is an empty [=/set=]:
+
+        1. |=list/Remove=| all |event names| from |subscription|'s event names.
+
+        1. If |subscription|'s event names is not empty:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+      1. Otherwise:
+
+        1. If |subscription| is [=subscription/global=]:
+
+          1. [=list/append=] |subscription| to |new subscriptions|.
+
+        1. Otherwise:
+
+          1. Let |event map| be an empty [=/map=].
+
+          1. For each |event name| in |subscription|'s event names:
+
+              1. Set |event map|[|event name|] to [=set/clone=] of |subscription|'s contexts.
+
+          1. For each |event name| in |event names|:
+
+              1. If |event map|[|event name|] does not exist,
+                 continue.
+
+              1. Remove |contexts| from |event map|[|event name|].
```

Comment:
Infra defines [remove](https://infra.spec.whatwg.org/#list-remove), but not as a set difference operation. You could probably write `Remove all items where |contexts| [=set/contains=] item from |event map|[|event name|].` although it's pretty awkward. I think we should consider adding a `difference` operation to infra so that we could write `Set |event map|[|event name|] to the [=set/difference=] between |event map|[|event name|] and |contexts|.`